### PR TITLE
[Compiler-V2] add constraints when type checking the cast operation

### LINKED ISSUE #11810: [Bug][move-compiler-v2] problem with lambda parameter type in conversion to move-model
# 🐛 Bug

The following example compiles on V1, but yields an error on V2:
```
module 0x12391283::M {
    use std::vector;
    fun main() : u64 {
        let gas_schedule_blob: vector<u8> = vector[
            115, 115, 95, 112, 97, 99, 107, 101, 100, 32, 0, 0, 0, 0, 0, 0, 0,
        ];
        vector::fold<u64, u8>(gas_schedule_blob, (0 as u64), |sum, addend| sum + (addend as u64))
    }
}
```
I see the error:
```
Diagnostics:
error: operand of cast must be a number
    ┌─ tests/folding/gas_schedule_constant.move:782
    │
  7│         vector::fold<u64, u8>(gas_schedule_blob, (0 as u64), |sum, addend| sum + (addend as u64))
    │                                                                                  ^^^^^^^^^^^^^^^
```
This can even run as a transactional test on V1, but it's failing in `exp_builder` on V2:
```

Diagnostics:
error: operand of cast must be a number
Backtrace: Backtrace [
    { fn: "std::backtrace_rs::backtrace::libunwind::trace", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/../../backtrace/src/backtrace/libunwind.rs", line: 93 },
    { fn: "std::backtrace_rs::backtrace::trace_unsynchronized", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/../../backtrace/src/backtrace/mod.rs", line: 66 },
    { fn: "std::backtrace::Backtrace::create", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/backtrace.rs", line: 331 },
    { fn: "move_model::model::GlobalEnv::add_backtrace", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/model.rs", line: 799 },
    { fn: "move_model::model::GlobalEnv::diag_with_notes", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/model.rs", line: 861 },
    { fn: "move_model::model::GlobalEnv::error_with_notes", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/model.rs", line: 827 },
    { fn: "move_model::builder::model_builder::ModelBuilder::error_with_notes", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/model_builder.rs", line: 224 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::error_with_notes", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 260 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::error", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 253 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_exp", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 1411 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_exp_free", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 2100 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_exp_list::{{closure}}", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 2812 },
    { fn: "core::iter::adapters::map::map_fold::{{closure}}", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/iter/adapters/map.rs", line: 84 },
    { fn: "<core::slice::iter::Iter<T> as core::iter::traits::iterator::Iterator>::fold", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/slice/iter/macros.rs", line: 232 },
    { fn: "<core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/iter/adapters/map.rs", line: 124 },
    { fn: "core::iter::traits::iterator::Iterator::for_each", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/iter/traits/iterator.rs", line: 857 },
    { fn: "alloc::vec::Vec<T,A>::extend_trusted", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/vec/mod.rs", line: 2881 },
    { fn: "<alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/vec/spec_extend.rs", line: 26 },
    { fn: "<alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/vec/spec_from_iter_nested.rs", line: 62 },
    { fn: "<alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/vec/spec_from_iter.rs", line: 33 },
    { fn: "<alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/vec/mod.rs", line: 2749 },
    { fn: "core::iter::traits::iterator::Iterator::collect", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/iter/traits/iterator.rs", line: 2053 },
    { fn: "itertools::Itertools::collect_vec", file: "/Users/brm/.cargo/registry/src/index.crates.io-6f17d22bba15001f/itertools-0.10.5/src/lib.rs", line: 1985 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_exp_list", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 2809 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_call", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 2545 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_exp", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 1286 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_lambda", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 3247 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_exp", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 1269 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_exp_free", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 2100 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_exp_list::{{closure}}", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 2812 },
    { fn: "core::iter::adapters::map::map_fold::{{closure}}", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/iter/adapters/map.rs", line: 84 },
    { fn: "<core::slice::iter::Iter<T> as core::iter::traits::iterator::Iterator>::fold", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/slice/iter/macros.rs", line: 232 },
    { fn: "<core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/iter/adapters/map.rs", line: 124 },
    { fn: "core::iter::traits::iterator::Iterator::for_each", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/iter/traits/iterator.rs", line: 857 },
    { fn: "alloc::vec::Vec<T,A>::extend_trusted", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/vec/mod.rs", line: 2881 },
    { fn: "<alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/vec/spec_extend.rs", line: 26 },
    { fn: "<alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/vec/spec_from_iter_nested.rs", line: 62 },
    { fn: "<alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/vec/spec_from_iter.rs", line: 33 },
    { fn: "<alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/vec/mod.rs", line: 2749 },
    { fn: "core::iter::traits::iterator::Iterator::collect", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/iter/traits/iterator.rs", line: 2053 },
    { fn: "itertools::Itertools::collect_vec", file: "/Users/brm/.cargo/registry/src/index.crates.io-6f17d22bba15001f/itertools-0.10.5/src/lib.rs", line: 1985 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_exp_list", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 2809 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_call", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 2545 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_fun_call", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 1921 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_exp", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 1200 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_seq_recursively", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 2223 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_seq_recursively", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 2165 },
    { fn: "move_model::builder::exp_builder::ExpTranslator::translate_seq", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/exp_builder.rs", line: 2112 },
    { fn: "move_model::builder::module_builder::ModuleBuilder::def_ana_fun::{{closure}}", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/module_builder.rs", line: 1370 },
    { fn: "move_model::builder::module_builder::ModuleBuilder::def_ana_fun", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/module_builder.rs", line: 1400 },
    { fn: "move_model::builder::module_builder::ModuleBuilder::def_ana", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/module_builder.rs", line: 890 },
    { fn: "move_model::builder::module_builder::ModuleBuilder::translate", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/builder/module_builder.rs", line: 201 },
    { fn: "move_model::run_move_checker", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/lib.rs", line: 378 },
    { fn: "move_model::run_model_builder_with_options_and_compilation_flags", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/lib.rs", line: 351 },
    { fn: "move_model::run_model_builder_in_compiler_mode", file: "/Users/brm/code/aptos-core/third_party/move/move-model/src/lib.rs", line: 93 },
    { fn: "move_compiler_v2::run_checker", file: "./src/lib.rs", line: 125 },
    { fn: "testsuite::TestConfig::run", file: "./tests/testsuite.rs", line: 285 },
    { fn: "testsuite::test_runner", file: "./tests/testsuite.rs", line: 82 },
    { fn: "datatest_stable::runner::Requirements::expand::{{closure}}::{{closure}}", file: "/Users/brm/.cargo/registry/src/index.crates.io-6f17d22bba15001f/datatest-stable-0.1.3/src/runner.rs", line: 59 },
    { fn: "libtest_mimic::Trial::test::{{closure}}", file: "/Users/brm/.cargo/registry/src/index.crates.io-6f17d22bba15001f/libtest-mimic-0.5.2/src/lib.rs", line: 110 },
    { fn: "core::ops::function::FnOnce::call_once{{vtable.shim}}", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/ops/function.rs", line: 250 },
    { fn: "<alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/boxed.rs", line: 2007 },
    { fn: "libtest_mimic::run_single::{{closure}}", file: "/Users/brm/.cargo/registry/src/index.crates.io-6f17d22bba15001f/libtest-mimic-0.5.2/src/lib.rs", line: 500 },
    { fn: "<core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/panic/unwind_safe.rs", line: 271 },
    { fn: "std::panicking::try::do_call", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs", line: 504 },
    { fn: "___rust_try" },
    { fn: "std::panicking::try", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs", line: 468 },
    { fn: "std::panic::catch_unwind", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panic.rs", line: 142 },
    { fn: "libtest_mimic::run_single", file: "/Users/brm/.cargo/registry/src/index.crates.io-6f17d22bba15001f/libtest-mimic-0.5.2/src/lib.rs", line: 500 },
    { fn: "libtest_mimic::run", file: "/Users/brm/.cargo/registry/src/index.crates.io-6f17d22bba15001f/libtest-mimic-0.5.2/src/lib.rs", line: 452 },
    { fn: "datatest_stable::runner::runner", file: "/Users/brm/.cargo/registry/src/index.crates.io-6f17d22bba15001f/datatest-stable-0.1.3/src/runner.rs", line: 17 },
    { fn: "testsuite::main", file: "/Users/brm/.cargo/registry/src/index.crates.io-6f17d22bba15001f/datatest-stable-0.1.3/src/macros.rs", line: 25 },
    { fn: "core::ops::function::FnOnce::call_once", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/ops/function.rs", line: 250 },
    { fn: "std::sys_common::backtrace::__rust_begin_short_backtrace", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/sys_common/backtrace.rs", line: 154 },
    { fn: "std::rt::lang_start::{{closure}}", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/rt.rs", line: 166 },
    { fn: "core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/ops/function.rs", line: 284 },
    { fn: "std::panicking::try::do_call", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs", line: 504 },
    { fn: "std::panicking::try", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs", line: 468 },
    { fn: "std::panic::catch_unwind", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panic.rs", line: 142 },
    { fn: "std::rt::lang_start_internal::{{closure}}", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/rt.rs", line: 148 },
    { fn: "std::panicking::try::do_call", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs", line: 504 },
    { fn: "std::panicking::try", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs", line: 468 },
    { fn: "std::panic::catch_unwind", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panic.rs", line: 142 },
    { fn: "std::rt::lang_start_internal", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/rt.rs", line: 148 },
    { fn: "std::rt::lang_start", file: "/rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/rt.rs", line: 165 },
    { fn: "_main" },
]
```

### PR DESCRIPTION
**Pull request by rahxephon89**
### Description

This PR closes #11810 by adding number type constraint to the expression type if it is not fully instantiated.

### Test Plan

Added two test cases to validate the fix.

### CODE REVIEW

File: third_party/move/move-compiler-v2/tests/checking/inlining/lambda_cast.move [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module 0x12391283::M {
+    use std::vector;
+    fun main(gas_schedule_blob: vector<u8>) : u64 {
+        let gas_schedule_blob: vector<u8> = vector[
+            115, 115, 95, 112, 97, 99, 107, 101, 100, 32, 0, 0, 0, 0, 0, 0, 0,
+        ];
+        vector::fold<u64, u8>(gas_schedule_blob, (0 as u64), |sum, addend| sum + (addend as u64))
```
Comment (brmataptos) [inline third_party/move/move-compiler-v2/tests/checking/inlining/lambda_cast.move:7]: What happens if we remove the explicit type parameter here?  `vector::fold(...)`

Also can we add a second test where this conversion is invalid to show that it generates an error?  For example, vector<vector<u8>>?
```
module 0x12391283::M {
    use std::vector;
    fun main(gas_schedule_blob: vector<u8>) : u64 {
        let gas_schedule_blob: vector<vector<u8>> = vector[
            vector[115], vector[115], vector[95], 
        ];
        vector::fold(gas_schedule_blob, (0 as u64), |sum, addend| sum + (addend as u64))
    }
}
```

**Review by brmataptos — COMMENTED** [review]
Need a few more test cases.

File: third_party/move/move-compiler-v2/tests/checking/inlining/lambda_cast.move (hunk 1 above)
Comment (rahxephon89) [reply third_party/move/move-compiler-v2/tests/checking/inlining/lambda_cast.move:7]: done

File: third_party/move/move-compiler-v2/tests/checking/inlining/lambda_cast_err.exp [hunk 2]
```diff
@@ -0,0 +1,7 @@
+
+Diagnostics:
+error: invalid call of `vector::fold`: expected `integer` but found `vector<u8>` for argument 3
```
Comment (brmataptos) [inline third_party/move/move-compiler-v2/tests/checking/inlining/lambda_cast_err.exp:3]: At least we have an error, but can you file a medium-priority bug to revisit this message?  The relationship between `vector<u8>` and `argument 3` is pretty indirect.

**Review by brmataptos — APPROVED** [review]
Good enough for now, but the error message needs some refinement eventually.

**Review by vineethk — APPROVED** [review]

**Review by wrwg — APPROVED** [review]

**Pull request merged**